### PR TITLE
STM32: Move RCC and system clocks configuration to device tree.

### LINKED ISSUE #17745: stm32: Move clock configuration to device tree
**Is your enhancement proposal related to a problem? Please describe.**
Current clock-control configuration (ie system clock source configuration) for stm32 based devices currently happens via Kconfig. To comply with overall device configuration, and follow other zephyr clock evolution (#15363, #16958) this should be moved to device tree.

Supersets #17703

**Describe the solution you'd like**
Bindings used for clock configuration should be similar to usual Linux clocks bindings.
Only difference would be new bindings introduced for pll parameters (various divisors, pre-divisors and multipliers) following current Kconfig scheme. This deviates from usual Linux binding which usually provides the output pll freq and use an algo at run time to calculate the various parameters allowing to set the required freq. At a later step, we might study the run time calculation option but for now, considering the range of clock-control IPs to study, keeping the current configuration scheme should speed up the implementation.
Same choice would be done for bus prescalers.

**Describe alternatives you've considered**

**Additional context**
Add any other context or graphics (drag-and-drop an image) about the feature request here.

### PR DESCRIPTION
**Pull request by erwango**
Provide initial and minimum changes to introduce device tree based clocks configuration
for STM32 based targets.

This initial PR extends to L4, WB and F4 series to ensure minimum scalability of the
proposed solution. Work on other series will be started once this RFC is approved.

Several requirements were taken into account:
- The solution should coexist with current Kconfig based configuration during the transition period
- Once transition is started, we should be able to generate deprecation warnings for Kconfig configuration 
out of tree users
- Minimize code changes
- Generate errors on erroneous device tree based clocks configuration

The proposed device tree clocks configuration is based on 2 parts:
- A clocks node that define all fixed clocks and plls.
- Additional properties given to rcc node to enable configuration of
  - System clock source selection (HSI, HE, PLL)
  - System clock-frequency
  - Bus configuration (APB, AHB, ..)

Fixes #17745

Note: Following compliance warnings are raised by compliance checks. I don't plan to fix them:
```
-:445: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
#445: FILE: drivers/clock_control/clock_stm32_ll_common.c:430:
+	if (IS_ENABLED(STM32_HSE_BYPASS)) {

-:466: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
#466: FILE: drivers/clock_control/clock_stm32_ll_common.c:477:
+		if (IS_ENABLED(STM32_HSE_BYPASS)) {
```

### CODE REVIEW

**Comment by erwango** [comment]
Reported warnings:
```
-:445: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
#445: FILE: drivers/clock_control/clock_stm32_ll_common.c:430:
+	if (IS_ENABLED(STM32_HSE_BYPASS)) {

-:466: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
#466: FILE: drivers/clock_control/clock_stm32_ll_common.c:477:
+		if (IS_ENABLED(STM32_HSE_BYPASS)) {
```

I don't plan to fix those for now.

**Review by galak — COMMENTED** [review]
At a quick glance at the bindings, it feels like we should see if we can represent the various clocks more fully in the devicetree.

Is there a particular SoC/family we can focus on to try and get this described cleanly and than apply that to the others?

**Comment by galak** [comment]
> Reported warnings:
> 
> ```
> -:445: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
> #445: FILE: drivers/clock_control/clock_stm32_ll_common.c:430:
> +	if (IS_ENABLED(STM32_HSE_BYPASS)) {
> 
> -:466: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
> #466: FILE: drivers/clock_control/clock_stm32_ll_common.c:477:
> +		if (IS_ENABLED(STM32_HSE_BYPASS)) {
> ```
> 
> I don't plan to fix those for now.
> 
> BTW: I wonder if we should keep `IS_ENABLED_CONFIG` warning in checkpatch.pl. Any opinion @galak ?

Seems fair to send a separate PR to disable that/remove it from  checkpatch for Zephyr.

File: include/drivers/clock_control/stm32_clock_control.h
Comment (galak) [inline include/drivers/clock_control/stm32_clock_control.h:18]: Add a comment block here about Kconfg v DTS

Comment (erwango) [reply include/drivers/clock_control/stm32_clock_control.h:30]: Done

**Comment by erwango** [comment]
> At a quick glance at the bindings, it feels like we should see if we can represent the various clocks more fully in the devicetree.

I think this is the case already. For instance on L4, we're introducing specific nodes for each potential input system clock.
Cf `stm32l4.dtsi`:
```
	clocks {
		clk_hse: clk-hse {
			#clock-cells = <0>;
			compatible = "st,stm32-hse-clock";
			status = "disabled";
		};

		clk_hsi: clk-hsi {
			#clock-cells = <0>;
			compatible = "fixed-clock";
			clock-frequency = <DT_FREQ_M(16)>;
			status = "disabled";
		};

		clk_msi: clk-msi {
			#clock-cells = <0>;
			compatible = "st,stm32-msi-clock";
			msi-range = <6>; /* 4MHz (reset value) */
			status = "disabled";
		};

		clk_lse: clk-lse {
			#clock-cells = <0>;
			compatible = "fixed-clock";
			clock-frequency = <32768>;
			status = "disabled";
		};

		clk_lsi: clk-lsi {
			#clock-cells = <0>;
			compatible = "fixed-clock";
			clock-frequency = <DT_FREQ_K(32)>;
			status = "disabled";
		};

		pll: pll {
			#clock-cells = <0>;
			compatible = "st,stm32l4-pll-clock";
			div-m = <1>;
			mul-n = <1>;
			div-p = <7>;
			div-q = <2>;
			div-r = <2>;
			/* clock-output-names = "pllclk", "pllsai3clk", "pll48m1clk"; */
			status = "disabled";
		};
	};
```
Each of these clocks could be configured/enabled/disabled to set up proper System Clock configuration.

File: dts/bindings/clock/st,stm32-rcc.yaml
Comment (gmarull) [inline dts/bindings/clock/st,stm32-rcc.yaml:20]: nit: `clock-frequency`

File: dts/bindings/clock/st,stm32f4-pll-clock.yaml
Comment (gmarull) [inline dts/bindings/clock/st,stm32f4-pll-clock.yaml:4]: It would be nice to document PLL "formulas", i.e. output frequency = f(input frequency, properties)

File: dts/arm/st/l4/stm32l4.dtsi
Comment (gmarull) [inline dts/arm/st/l4/stm32l4.dtsi:80]: Why is this property commented out?

File: boards/arm/nucleo_l476rg/nucleo_l476rg.dts
Comment (gmarull) [inline boards/arm/nucleo_l476rg/nucleo_l476rg.dts:80]: I assume this value could be automatically computed given HSI clock value is known as well as PLL formulae. Could be done in the future, though. I'm not sure, though, if it's in general better to specify `clock-frequency` and let the driver figure out the PLL values. Again, likely for future revisions.

File: drivers/clock_control/Kconfig.stm32
Comment (gmarull) [inline drivers/clock_control/Kconfig.stm32:78]: nit: missing indentation

**Review by gmarull — COMMENTED** [review]
Changes and approach LGTM, specially keeping compatibility with Kconfig for now (transition should be smooth).

File: dts/bindings/clock/st,stm32-rcc.yaml
Comment (erwango) [reply dts/bindings/clock/st,stm32-rcc.yaml:20]: txs

File: dts/bindings/clock/st,stm32f4-pll-clock.yaml
Comment (erwango) [reply dts/bindings/clock/st,stm32f4-pll-clock.yaml:4]: Right, this will be clearer! Adding this.

File: dts/arm/st/l4/stm32l4.dtsi
Comment (erwango) [reply dts/arm/st/l4/stm32l4.dtsi:80]: We're not ready to support that yet.
I'll remove

File: boards/arm/nucleo_l476rg/nucleo_l476rg.dts
Comment (erwango) [reply boards/arm/nucleo_l476rg/nucleo_l476rg.dts:80]: Yes, I don't see obstacles to get it computed automatically, but for this step I prefer to keep same level of features.

File: drivers/clock_control/Kconfig.stm32
Comment (erwango) [reply drivers/clock_control/Kconfig.stm32:78]: It was initially indented but I got errors reported by Kconfig scripts.
I will have a new try but I'm not confident.

**Comment by erwango** [comment]
@gmarull Thanks for the review. I've added the PLL description

**Review by gmarull — APPROVED** [review]

File: dts/bindings/clock/st,stm32f4-pll-clock.yaml
Comment (martinjaeger) [inline dts/bindings/clock/st,stm32f4-pll-clock.yaml:52]: This is required, otherwise the driver will not compile (to verify comment it out in board dts file and run `west build -b nucleo_f429zi samples/basic/blinky -p`). 

Could be the same for `div-p` and `div-q` of other MCU series. I was not sure why just those two divisors were declared as non-required in all bindings. Was this on purpose?

File: dts/bindings/clock/st,stm32-rcc.yaml
Comment (martinjaeger) [inline dts/bindings/clock/st,stm32-rcc.yaml:15]: Nits:
```suggestion
  Last bus clocks (typically HCLK, PCLK1, PCLK2) should be configured using matching prescaler
```

File: dts/bindings/clock/st,stm32-rcc.yaml
Comment (martinjaeger) [inline dts/bindings/clock/st,stm32-rcc.yaml:17]: Nits:
```suggestion
  Here is an example of correctly configured rcc node:
```

File: dts/bindings/clock/st,stm32-rcc.yaml
Comment (martinjaeger) [inline dts/bindings/clock/st,stm32-rcc.yaml:37]: Typo:
```suggestion
  control the peripheral clock in that bus register.
```

File: dts/common/freq.h
Comment (martinjaeger) [inline dts/common/freq.h:11]: Why not use existing macros `KHZ()` and `MHZ()` from `include/sys/util.h`?

File: include/drivers/clock_control/stm32_clock_control.h
Comment (martinjaeger) [inline include/drivers/clock_control/stm32_clock_control.h:26]: Nits:
```suggestion
 * Finally, once all in-tree users are converted to dts based configuration,
```

**Review by martinjaeger — COMMENTED** [review]
Really great to get the clock configuration into devicetree.

File: dts/bindings/clock/st,stm32f4-pll-clock.yaml
Comment (erwango) [reply dts/bindings/clock/st,stm32f4-pll-clock.yaml:52]: There are 2 things to consider:

#### In the cock driver implementation:
The required div depends on implementation:
-F4: Only div-P is required. Q is optional
-L4/WB: Only div-R is required. Q is optional

#### Considering the bindings:
So far (in the series I have worked on), on a series, various PLLs are made out of the same hardware block (ie various factors support the same ranges).
So, in order to limit the number of bindings files to be created, I tried to take advantage of that and used only one compatible to describe all PLL of SoC. For instance, compatible `st,stm32l4-pll-clock` could be used to describe PLL, PLLSAI1 and PLLSAI2 (which is purely theoretical for now as only the main one being implemented in clock driver).

On F4 the same rule applies, but there's a trick: PLL are identical, but don't use the same output:
Main PLL uses output P (-> PLLCLK) and output Q (-> PLL48CLK)
PLLI2S uses output R (PLLI2SCLK)
So, instead of creating 2 compatibles that would be extremely close, I've created one compatible with optional (/not required) outputs.
Alternative would be to create `st,stm32f4-plli2s-clock`, inheriting from `st,stm32f4-pll-clock`, deleting div-p and div-q and  adding div-r, and setting all as `required: true`. This is an interesting option: more "binding".

This whole summary helped to change my mind in favor of this last option, as I think this should help minimizing users errors. Thanks for this question !

File: dts/common/freq.h
Comment (erwango) [reply dts/common/freq.h:11]: My understanding is that dts files should remain independent from zephyr code.

**Comment by erwango** [comment]
@martinjaeger Thanks for the review, I've made the appropriate changes

File: dts/common/freq.h
Comment (martinjaeger) [reply dts/common/freq.h:11]: Ok, makes sense.

I would probably prefer to have the entire unit (e.g. `DT_FREQ_KHZ` or just `DT_KHZ`) in the name, but won't insist.

**Review by martinjaeger — APPROVED** [review]

File: dts/bindings/clock/st,stm32f4-pll-clock.yaml
Comment (martinjaeger) [reply dts/bindings/clock/st,stm32f4-pll-clock.yaml]: Thanks for the explanation. Honestly I'm not familiar with the differences between the series in that much detail. But it sounds reasonable to duplicate a bit of the PLL code in order to make sure it matches more exactly to a particular series.

File: drivers/clock_control/clock_stm32f0_f3.c
Comment (FRASTM) [inline drivers/clock_control/clock_stm32f0_f3.c:19]: `CONFIG_CLOCK_` prefix was removed in previous changes, why a `CONFIG_SOC_` prefix here ?

Comment (erwango) [reply drivers/clock_control/clock_stm32f0_f3.c:19]: Indeed. Since I'm not changing F0/F3 (and L0/L1 ...), I've preferred to limit the changes in those files and reverted few of the changes. They will be done in later PR once this one is approved.

Comment (FRASTM) [reply drivers/clock_control/clock_stm32f0_f3.c:19]: ok

**Review by FRASTM — APPROVED** [review]

**Comment by erwango** [comment]
Last push to fix `hse-bypass` selection.

File: dts/bindings/clock/st,stm32-msi-clock.yaml
Comment (mbolivar-nordic) [inline dts/bindings/clock/st,stm32-msi-clock.yaml:28]: No `enum`?

File: dts/bindings/clock/st,stm32-rcc.yaml
Comment (mbolivar-nordic) [inline dts/bindings/clock/st,stm32-rcc.yaml:60]: This is a bit weird; do you really need to have this property now in this way?

File: dts/bindings/clock/st,stm32wb-rcc.yaml
Comment (mbolivar-nordic) [inline dts/bindings/clock/st,stm32wb-rcc.yaml:20]: Same question here about whether this property is needed yet.

File: dts/common/freq.h
Comment (mbolivar-nordic) [reply dts/common/freq.h:11]: > My understanding is that dts files should remain independent from zephyr code.

util_macro.h is available from DTS; see dt-util.h. Perhaps we need to move some definitions into it?

**Review by mbolivar-nordic — COMMENTED** [review]
Very cool. A couple of questions.

File: dts/bindings/clock/st,stm32-msi-clock.yaml
Comment (erwango) [reply dts/bindings/clock/st,stm32-msi-clock.yaml:28]: enum would work indeed. I'll make the change.

File: dts/bindings/clock/st,stm32-rcc.yaml
Comment (erwango) [reply dts/bindings/clock/st,stm32-rcc.yaml:60]: I agree this is a bit weird but I wanted this property to be present for 2 reasons:
- backward compatibility with existing scheme
- this is a typical STM32 configuration piece. We have a tool to help users configure clocks and this is part of the tool output. So, despite the current constraint on this field in Zephyr, I don't want to hide it in any way in order not to disturb users and to let them inform the earliest of this fact.

Aside the weirdness, is there any concern using 'enum' this way ?

File: dts/bindings/clock/st,stm32wb-rcc.yaml
Comment (erwango) [reply dts/bindings/clock/st,stm32wb-rcc.yaml:20]: Same answer as this is the exact same `ahb-prescaler` behind the scene.

**Comment by erwango** [comment]
@mbolivar-nordic, txs for the review. I've converted `msi-range` to enum and answered about the 2 'weird' fields.
About

> util_macro.h is available from DTS; see dt-util.h. Perhaps we need to move some definitions into it?

I think this should be the same also for existing mem.h (which defines `DT_SIZE_K` and such).
So I think they should be handled the same way. Also whereever we put this, I propose we get the 
file included by `skeleton.dtsi` so we don't have to deal with it anymore.

File: dts/bindings/clock/st,stm32-rcc.yaml
Comment (mbolivar-nordic) [reply dts/bindings/clock/st,stm32-rcc.yaml:60]: Makes sense, thanks

**Comment by erwango** [comment]
> I think this should be the same also for existing mem.h (which defines DT_SIZE_K and such).
So I think they should be handled the same way. Also whereever we put this, I propose we get the
file included by skeleton.dtsi so we don't have to deal with it anymore.

@galak. Seems that this is the only remaining open point on the PR.
I propose to do the change in a follow up PR once we agree on the file it should belong to (you're input is welcome on that part).

**Comment by galak** [comment]
While not 100% related, but do we need `CONFIG_CLOCK_CONTROL=y` in the defconfig (for example boards/arm/nucleo_wb55rg/nucleo_wb55rg_defconfig).  Seems like nothing w/function without this being enabled.

**Review by galak — APPROVED** [review]

**Comment by erwango** [comment]
> While not 100% related, but do we need CONFIG_CLOCK_CONTROL=y in the defconfig (for example boards/arm/nucleo_wb55rg/nucleo_wb55rg_defconfig). Seems like nothing w/function without this being enabled.

I had the same remark. I'll remove once all boards are converted (even if not strictly dependent could be a nice cleaning follow up action).

**Comment by galak** [comment]
> I had the same remark. I'll remove once all boards are converted (even if not strictly dependent could be a nice cleaning follow up action).

Yeah, was going to say to deal with it as a follow on commit.  (hoping to see one day if we get to the point that there is no board defconfig :).

**Comment by erwango** [comment]
Seems BK is down: 
```

s3secrets-helper: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by s3secrets-helper)
--
  | 🚨 Error: Error setting up bootstrap: The global environment hook exited with status 1


```

**Comment by galak** [comment]
> Seems BK is down:
> 
> ```
> 
> s3secrets-helper: /lib64/libc.so.6: version `GLIBC_2.32' not found (required by s3secrets-helper)
> --
>   | 🚨 Error: Error setting up bootstrap: The global environment hook exited with status 1
> ```

Should be fixed.  I kicked off the PR to build again.

**Comment by erwango** [comment]
> Should be fixed. I kicked off the PR to build again.

Thanks. There was a rebase issue which is now flxed as well.

**Comment by galak** [comment]
@erwango is this still RFC?

**Comment by erwango** [comment]
@galak 'RFC' removed

**Comment by galak** [comment]
@erwango any comments about the compliance issues.  Some of these some real like CODEOWNERS for dts/common/freq.h (you can put me down if you want), CONFIG_SOC_STM32_PLL_SRC_PLL2 and CONFIG_SOC_STM32_SYSCLK_SRC_PLL no longer existing, but still having refs in the code.

The IS_ENABLED seems a bit more tricky.

**Comment by erwango** [comment]
> @erwango any comments about the compliance issues. Some of these some real like CODEOWNERS for dts/common/freq.h (you can put me down if you want), 

@galak can you elaborate a bit ?

> CONFIG_SOC_STM32_PLL_SRC_PLL2 

This one can have usage in out or tree code and we have plans to put real usage in tree.

> CONFIG_SOC_STM32_SYSCLK_SRC_PLL 

This one exists and is used.

> The IS_ENABLED seems a bit more tricky

`IS_ENABLED` has usages w/o CONFIG_ symbols today (some `IS_ENABLED(DT_FOO)` could be found).

**Comment by galak** [comment]
> > 
> 
> @galak can you elaborate a bit ? Current proposal is based on mem.h. But I can move them both to another file after if needed. Just need to know the file.

the file is not listed in CODEOWNERS as what I believe the issue is.

**Comment by galak** [comment]
> > CONFIG_SOC_STM32_PLL_SRC_PLL2
> 
> This one can have usage in out or tree code and we have plans to put real usage in tree.
> 
> > CONFIG_SOC_STM32_SYSCLK_SRC_PLL
> 
> This one exists and is used.

Where are the Kconfig symbols associated with these?

**Comment by erwango** [comment]
> > > CONFIG_SOC_STM32_PLL_SRC_PLL2
> > 
> > 
> > This one can have usage in out or tree code and we have plans to put real usage in tree.
> > > CONFIG_SOC_STM32_SYSCLK_SRC_PLL
> > 
> > 
> > This one exists and is used.
> 
> Where are the Kconfig symbols associated with these?

Right, gotcha. Indeed those don't exist and should be cleaned up (Should be `CLOCK_STM32_...` instead of `SOC_STM32_...`)
I've a fixed up patch in the up comping second part of this change (when converting the respective impacted series: L0/L1/F0/F1/F3).

**Comment by galak** [comment]
> > 
> 
> Right, gotcha. Indeed those don't exist and should be cleaned up (Should be `CLOCK_STM32_...` instead of `SOC_STM32_...`)
> I'm cleaning this up in comping second part of this change (when converting the respective impacted series: L0/L1/F0/F1/F3).

confused, are you saying the cleanup to remove the no-existent kconfig symbols in a follow up PR?

**Comment by erwango** [comment]
@galak CI timed-out, otherwise looks good.

**Comment by carlescufi** [comment]
@erwango compliance is failing due to missing Kconfig symbols, this one should be fixed.

> > > 
> > 
> > 
> > Right, gotcha. Indeed those don't exist and should be cleaned up (Should be `CLOCK_STM32_...` instead of `SOC_STM32_...`)
> > I'm cleaning this up in comping second part of this change (when converting the respective impacted series: L0/L1/F0/F1/F3).
> 
> confused, are you saying the cleanup to remove the no-existent kconfig symbols in a follow up PR?

Agreed, I think we should fix those missing Kconfig in this PR

**Comment by erwango** [comment]
@carlescufi @galak Kconfig symbol fixed. Hope this is fine now.

**Pull request merged**
